File: core/src/main/java/me/raatiniemi/worker/domain/model/TimeIntervalStartingPoint.java
/*
 * Copyright (C) 2018 Worker Project
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package me.raatiniemi.worker.domain.model;

import java.util.Calendar;

import me.raatiniemi.worker.domain.exception.InvalidStartingPointException;

public final class TimeIntervalStartingPoint {
    public static final int DAY = 0;
    public static final int WEEK = 1;
    public static final int MONTH = 2;

    private TimeIntervalStartingPoint() {
    }

    public static long getMillisecondsForStartingPoint(int startingPoint) {
        Calendar calendar = Calendar.getInstance();
        calendar.set(Calendar.HOUR_OF_DAY, 0);
        calendar.set(Calendar.MINUTE, 0);
        calendar.set(Calendar.SECOND, 0);
        calendar.set(Calendar.MILLISECOND, 0);

        switch (startingPoint) {
            case DAY:
                break;
            case WEEK:
                calendar.setFirstDayOfWeek(Calendar.MONDAY);
                calendar.set(Calendar.DAY_OF_WEEK, Calendar.MONDAY);
                break;
            case MONTH:
                calendar.set(Calendar.DAY_OF_MONTH, 1);
                break;
            default:
                throw new InvalidStartingPointException(
                        "Starting point '" + startingPoint + "' is not valid"
                );
        }

        return calendar.getTimeInMillis();
    }
}

File: core/src/main/java/me/raatiniemi/worker/domain/timeinterval/model/timeIntervalStartingPoint.kt
/*
 * Copyright (C) 2020 Tobias Raatiniemi
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package me.raatiniemi.worker.domain.timeinterval.model

import me.raatiniemi.worker.domain.configuration.AppKeys
import me.raatiniemi.worker.domain.configuration.KeyValueStore
import me.raatiniemi.worker.domain.timeinterval.usecase.InvalidStartingPointException

fun timeIntervalStartingPoint(keyValueStore: KeyValueStore): TimeIntervalStartingPoint {
    val defaultValue = TimeIntervalStartingPoint.MONTH
    val startingPoint = keyValueStore.int(
        AppKeys.TIME_SUMMARY,
        defaultValue.rawValue
    )

    return try {
        timeIntervalStartingPoint(startingPoint)
    } catch (e: InvalidStartingPointException) {
        defaultValue
    }
}

fun timeIntervalStartingPoint(startingPoint: Int): TimeIntervalStartingPoint {
    return when (startingPoint) {
        TimeIntervalStartingPoint.DAY.rawValue -> TimeIntervalStartingPoint.DAY
        TimeIntervalStartingPoint.WEEK.rawValue -> TimeIntervalStartingPoint.WEEK
        TimeIntervalStartingPoint.MONTH.rawValue -> TimeIntervalStartingPoint.MONTH
        else -> throw InvalidStartingPointException(
            "Starting point '$startingPoint' is not valid"
        )
    }
}
